src/runtime/timer.h

```diff
@@ -15,53 +30,95 @@ static inline void runloop_timer(timestamp duration)
     apply(platform_timer, duration);
 }
 
-timer register_timer(timestamp interval, clock_id id, thunk n);
-timer register_periodic_timer(timestamp interval, clock_id id, thunk n);
-timestamp remove_timer(timer t);
-void initialize_timers(kernel_heaps kh);
-timestamp parse_time();
-void print_timestamp(buffer, timestamp);
-timestamp timer_check();
+timer register_timer(clock_id id, timestamp val, boolean absolute, timestamp interval, timer_handler n);
 
-#define THOUSAND (1000ull)
-#define MILLION (1000000ull)
-#define BILLION (1000000000ull)
-#define QUADRILLION (1000000000000000ull)
-#define TIMESTAMP_SECOND (1ull << 32)
+#if defined(STAGE3) || defined(BUILD_VDSO)
+#include <vdso.h>
+#define __rtc_offset (&(VVAR_REF(vdso_dat)))->rtc_offset
+#else
+#define __rtc_offset 0
+#endif
 
-// danger - truncation, should always be subsec
-
-static inline timestamp seconds(u64 n)
+/* Convert to monotonic. Not clear yet how to map process and thread
+   times to monotonic scale. Should the process have its own timer heap? */
+static inline timestamp timer_expiry(timer t)
 {
-    return n * TIMESTAMP_SECOND;
+    switch (t->id) {
```

Comment:
right. sorry, this is really the block that seems like it should be in 'unix'

Reply:
I guess I started off assuming that these various clocksources would need sorting out at the lowest levels, so I made the internal CLOCK_ID_* identifiers line up with the unix CLOCK_* ones. But perhaps that assumption is wrong, as it seems in the end to just boil down to monotonic or realtime. So I guess we could revert to now() with no parameter and expose rtc_offset to the unix world. I supposed this could be addressed when implementing CLOCK_PROCESS_CPUTIME_ID and CLOCK_THREAD_CPUTIME_ID.